services/src/handlers/datasets.rs

```diff
@@ -97,8 +139,189 @@ fn adjust_user_path_to_upload_path(meta: &mut MetaDataDefinition, upload: &Uploa
+
+    let main_file_path = upload.id.root_path()?.join(&create.main_file);
+    let meta_data = auto_detect_dataset(&main_file_path)?;
+
+    let properties = AddDataSet {
+        id: None,
+        name: create.dataset_name,
+        description: create.dataset_description,
+        source_operator: "OgrSource".to_owned(),
+    };
+
+    let mut db = ctx.data_set_db_ref_mut().await;
+    let meta_data = db.wrap_meta_data(meta_data);
+    let id = db
+        .add_data_set(session.user.id, properties.validated()?, meta_data)
+        .await?;
+
+    Ok(warp::reply::json(&IdResponse::from(id)))
+}
+
+fn auto_detect_dataset(main_file_path: &Path) -> Result<MetaDataDefinition> {
+    let mut dataset = Dataset::open(&main_file_path).context(error::Gdal)?;
+    let layer = {
+        if let Ok(layer) = dataset.layer(0) {
+            layer
+        } else {
+            // TODO: handle Raster datasets as well
+            return Err(crate::error::Error::DataSetHasNoAutoImportableLayer);
+        }
+    };
+    let vector_type = detect_vector_type(&layer)?;
+    let spatial_reference: SpatialReference = layer
+        .spatial_ref()
+        .context(error::Gdal)?
+        .try_into()
+        .context(error::DataType)?;
+    let columns_map = detect_columns(&layer);
+    let columns_vecs = column_map_to_column_vecs(&columns_map);
+
+    Ok(MetaDataDefinition::OgrMetaData(StaticMetaData {
+        loading_info: OgrSourceDataset {
+            file_name: main_file_path.into(),
+            layer_name: layer.name(),
+            data_type: Some(vector_type),
+            time: OgrSourceDatasetTimeType::None, // TODO: auto detect time type and corresponding columns
+            columns: Some(OgrSourceColumnSpec {
+                x: "".to_owned(), // TODO: for csv-files: try to find wkt/xy columns
+                y: None,
+                numeric: columns_vecs.numeric,
+                decimal: columns_vecs.decimal,
+                textual: columns_vecs.textual,
+            }),
+            default_geometry: None,
+            force_ogr_time_filter: false,
+            on_error: geoengine_operators::source::OgrSourceErrorSpec::Skip,
+            provenance: None,
+        },
+        result_descriptor: VectorResultDescriptor {
+            data_type: vector_type,
+            spatial_reference: spatial_reference.into(),
+            columns: columns_map,
+        },
+    }))
+}
+
+fn detect_vector_type(layer: &Layer) -> Result<VectorDataType> {
+    let ogr_type = layer
+        .feature(0)
+        .ok_or(error::Error::EmptyDataSetCannotBeImported)?
+        .geometry()
+        .geometry_type();
+
+    VectorDataType::try_from_ogr_type_code(ogr_type).context(error::DataType)
+}
+
+struct Columns {
+    decimal: Vec<String>,
+    numeric: Vec<String>,
+    textual: Vec<String>,
+}
+
+fn detect_columns(layer: &Layer) -> HashMap<String, FeatureDataType> {
+    let mut columns = HashMap::default();
+    // TODO: avoid reading ALL of the data
```

Comment:
Couldn't you use the [`defn`](https://docs.rs/gdal/0.7.2/gdal/vector/struct.Layer.html#method.defn) method that returns  [`Defn`](https://docs.rs/gdal/0.7.2/gdal/vector/struct.Defn.html), which has a method [`fields`](https://docs.rs/gdal/0.7.2/gdal/vector/struct.Defn.html#method.fields)?

Reply:
that is a hard to find method 😅 yes, makes absolutely sense c8e6fe123441d3cfe993329a1edf7663afe85fbd